apps/services/search-indexer/src/migrate/lib/aws.ts

```diff
@@ -190,31 +216,64 @@ const uploadFileToS3 = (options: Omit<PutObjectRequest, 'Bucket'>) => {
     Bucket: params.Bucket,
     Key: params.Key,
   })
-  return s3
-    .upload(params)
-    .promise()
-    .catch((error) => {
-      logger.error('Failed to upload s3 package', error)
-      throw error
-    })
+  try {
+    return await s3.send(new PutObjectCommand(params))
+  } catch (error) {
+    logger.error('Failed to upload s3 package', error)
+    throw error
+  }
 }
 
-const checkIfS3Exists = (key: string) => {
-  return s3
-    .headObject({
-      Bucket: environment.s3Bucket,
-      Key: key,
-    })
-    .promise()
-    .then(() => true)
-    .catch(() => false) // we don't want this to throw on error so we can handle if not exists manually
+const checkIfS3Exists = async (key: string) => {
+  try {
+    await s3.send(
+      new HeadObjectCommand({
+        Bucket: environment.s3Bucket,
+        Key: key,
+      }),
+    )
+    return true
+  } catch {
+    return false
+  } // we don't want this to throw on error so we can handle if not exists manually
 }
 
 interface S3DictionaryFile {
   locale: ElasticsearchIndexLocale
   analyzerType: string
   version: string
 }
+
+export const streamToBuffer = async (
+  file: NodeJS.ReadableStream,
+): Promise<Buffer> => {
+  return new Promise((resolve, reject) => {
+    if (!(file instanceof Readable)) {
+      // In case the input is not a Readable stream (shouldn't happen due to typing)
+      reject(new Error('Input is not a valid Readable stream'))
+    }
+    const chunks: Uint8Array[] = []
+    file.on('data', (chunk) => chunks.push(chunk))
+    file.on('end', () => {
+      if (chunks.every((chunk) => typeof chunk === 'string')) {
+        // If all chunks are strings, concatenate them and convert to Buffer
+        resolve(Buffer.from(chunks.join('')))
+      } else if (
+        chunks.every(
+          (chunk) => Buffer.isBuffer(chunk) || chunk instanceof Uint8Array,
+        )
+      ) {
+        // If all chunks are Buffer or Uint8Array, use Buffer.concat
+        resolve(Buffer.concat(chunks))
```

Comment:
I'm not sure how big the buffers are going to be but do you think it is worth considering iterating through the chunks once, and then deciding on a resolve strategy?

Dirty pseudocode:
```pseudocode
declare x;
for every chunk {
  x["string"] += 1 if chunk is string
  x["isBuffer"] += 1 if chunk.isBuffer() or chunk is instanceof u8[]
};

..do the if else if else dance \O/
```

Reply:
Actually, nevermind. Thinking more on this. If we're frequently dealing with mixed types then exhausting the chunks once would likely run slower than the current solution.

Reply:
Marking this as resolved